Introduce DEAL_II_VECTORIZATION_WIDTH_IN_BITS

kronbichler opened the pull request:
Based on the discussion in #7321, this PR changes the unintuitive `DEAL_II_COMPILER_VECTORIZATION_LEVEL` with `DEAL_II_HIGHEST_N_VECTORIZATION_BITS`. I think this name is a good description of what happens; I do not think it makes sense to split doubles and floats because all systems where we support vectorization have the same number of bits for doubles and floats, and bits is also the marketing name for it rather than "length" where one could be confused by the number of doubles in the SIMD array versus bits. Opinions? Once we agree on the name, I will add a changelog in the incompatibilities section.

Fixes #7321.

tamiko commented on include/deal.II/base/config.h.in:
```diff
@@ -116,7 +116,21 @@
+#define DEAL_II_COMPILER_VECTORIZATION_LEVEL 1
+#else
+#define DEAL_II_COMPILER_VECTORIZATION_LEVEL 0
+#endif
```
Would you mind to test quickly whether it is possible to write something like
```
#if  DEAL_II_COMPILER_HAS_DIAGNOSTIC_PRAGMA
  // ...
  # define DEAL_II_COMPILER_VECTORIZATION_LEVEL _Pragma ("GCC warning \"The DEAL_II_COMPILER_VECTORIZATION_LEVEL macro is deprecated\"") 3
  // ...
#else 
  // what you have  
#endif 
```

kronbichler replied:
I will try that now.

kronbichler replied:
Works like a charm, I will add this together with the renaming of the variable.

bangerth commented on include/deal.II/base/config.h.in:
```diff
@@ -116,7 +116,21 @@
 
 #cmakedefine DEAL_II_WORDS_BIGENDIAN
-#define DEAL_II_COMPILER_VECTORIZATION_LEVEL @DEAL_II_COMPILER_VECTORIZATION_LEVEL@
+#define DEAL_II_HIGHEST_N_VECTORIZATION_BITS @DEAL_II_HIGHEST_N_VECTORIZATION_BITS@
```
Would you mind adding a comment here documenting what the macro is supposed to be represent (or where else to find documentation)?

tjhei commented:
How about something like ``VECTORIZATION_WIDTH`` or ``VECTORIZATION_WIDTH_IN_BITS``?

kronbichler commented:
I like `DEAL_II_VECTORIZATION_WIDTH_IN_BITS` - the only question is whether we insert a `LARGEST` or `MAX` before that because we do technically support narrower SIMD arrays as well, it's just that we need a flag for the widest. But I think this could be covered by comments at the place where we set the macro.

tjhei commented:
> the only question is whether we insert a `LARGEST` or `MAX` before that because we do technically support narrower SIMD arrays as well

I don't think that is necessary, especially for anyone familiar with hardware.

tjhei commented on cmake/configure/configure_vectorization.cmake:
```diff
@@ -21,19 +21,19 @@
    SET(DEAL_II_EXPAND_FLOAT_VECTORIZED  "${DEAL_II_EXPAND_FLOAT_VECTORIZED}" "VectorizedArray<float,4>")
 ENDIF()
 
-IF((${DEAL_II_COMPILER_VECTORIZATION_LEVEL} GREATER 1) AND ( DEAL_II_HAVE_AVX OR DEAL_II_HAVE_ALTIVEC))
```
what happened with altivec here?

kronbichler replied:
AltiVec is 128 bits, so the condition cannot be true, and the altivec case goes to the `VECTORIZATION_WIDTH_IN_BITS > 0` case. It took me a while to understand the condition, but I think we simply missed it; in any case, I think we could also skip the `AVX` and `AVX512` parts here because the macro tells us what we have.

tjhei commented on include/deal.II/base/utilities.h:
```diff
@@ -812,13 +812,13 @@ namespace Utilities
      * <table>
      * <tr>
-     *   <td><tt>VECTORIZATION_LEVEL</tt></td>
+     *   <td><tt>N_VECTORIZATION_BITS</tt></td>
```
do you want to update this caption?

kronbichler replied:
I did, because I do not think the `VECTORIZATION_LEVEL` makes much sense now any more. I was a bit unsure if we should completely remove this column now that we have the name `VECTORIZATION_WIDTH_IN_BITS` and two columns with the same information.

tjhei replied:
Sorry for my cryptic comment: I think you need to update this column name to reflect the rename to ``VECTORIZATION_WIDTH``.

kronbichler replied:
OK, I'll undo this change for now.

tjhei commented on include/deal.II/base/numbers.h:
```diff
@@ -81,13 +81,13 @@ namespace internal
+#if DEAL_II_VECTORIZATION_WIDTH_IN_BITS >= 128 && defined(__ALTIVEC__)
       2;
-#elif DEAL_II_COMPILER_VECTORIZATION_LEVEL >= 3 && defined(__AVX512F__)
+#elif DEAL_II_VECTORIZATION_WIDTH_IN_BITS >= 512 && defined(__AVX512F__)
```
are these ``&& defined(bla)`` necessary? I thought we catch this in vectorization.h

kronbichler replied:
You are right, we should be getting into these errors here: https://github.com/dealii/dealii/blob/61a022c0784163310939baa4dbde83971b89faa7/include/deal.II/base/vectorization.h#L46-L55
So I will simply remove the second part.

tjhei commented on include/deal.II/base/numbers.h:
```diff
@@ -81,13 +81,13 @@ namespace internal
      */
     constexpr static unsigned int max_width =
-#if DEAL_II_COMPILER_VECTORIZATION_LEVEL >= 1 && defined(__ALTIVEC__)
+#if DEAL_II_VECTORIZATION_WIDTH_IN_BITS >= 128 && defined(__ALTIVEC__)
```
why is altivec a special case here? Can't we just use the width and sort all entries here?

tjhei commented on include/deal.II/base/config.h.in:
```diff
@@ -116,7 +116,40 @@
+ */
+#ifdef DEAL_II_COMPILER_HAS_DIAGNOSTIC_PRAGMA
+#  if DEAL_II_VECTORIZATION_WIDTH_IN_BITS == 512
+#  define DEAL_II_COMPILER_VECTORIZATION_LEVEL _Pragma ("GCC warning \"The DEAL_II_COMPILER_VECTORIZATION_LEVEL macro is deprecated\"") 3
```
I don' think there is any harm in keeping this macro around (sorry, you already put some effort into this).

tjhei replied:
Same, I don't feel strongly about it. It just means getting annoying warnings in every code using matrix-free.

tjhei commented on cmake/checks/check_01_cpu_features.cmake:
```diff
@@ -277,17 +277,17 @@ ENDIF() # IF DEAL_II_ALLOW_PLATFORM_INTROSPECTION
 
 IF(DEAL_II_HAVE_ALTIVEC)
-  SET(DEAL_II_COMPILER_VECTORIZATION_LEVEL 1)
+  SET(DEAL_II_VECTORIZATION_WIDTH_IN_BITS 1)
```
why is this not 128?

tjhei commented on cmake/checks/check_01_cpu_features.cmake:
```diff
@@ -297,7 +297,7 @@ ENDIF()
 
 IF(DEAL_II_WITH_CUDA)
-  SET(DEAL_II_COMPILER_VECTORIZATION_LEVEL 0)
+  SET(DEAL_II_VECTORIZATION_WIDTH_IN_BITS 0)
```
unrelated to this PR but does this mean that CPU vectorization is disabled when compiling with CUDA support?

kronbichler replied:
Yes, we had problems with the CUDA compiler when we had intrinsics on that were difficult to resolve, see #7655 and #7542.

tjhei commented on include/deal.II/base/vectorization.h:
```diff
@@ -26,29 +26,29 @@
+#if DEAL_II_VECTORIZATION_WIDTH_IN_BITS > 0
 
-#  if DEAL_II_COMPILER_VECTORIZATION_LEVEL >= 2 && defined(__SSE2__) && \
+#  if DEAL_II_VECTORIZATION_WIDTH_IN_BITS >= 256 && defined(__SSE2__) && \
```
I don't understand why you need to check for the presence of SSE2 macro here. I know, that was here before...

kronbichler replied:
The problem we try to address here is mostly user projects that set the CPU flags on their own. What can happen is that deal.II is compiled with say AVX512, but then a user project forgets to add `-march=native` and falls back to SSE2. This leads to very strange errors as the size of data structures differs between the compiled deal.II code sitting in `libdeal_II.so` and the user code.
The second thing this is supposed to prevent is the case when switching between different CPUs with `-march=native`. One could blame the user for it, but I like to speak out when something goes wrong. I assume we should put this discussion in the file, right?

kronbichler replied:
Ah, but you mean specifically the SSE2 macro here, given that x86-64 is the only instruction set right now with 256 bits or more? I think the reasoning for using it was to prevent additional trouble when we have another instruction set (like ARM's SVE) where we could have 256 bits (or 512) but not defined AVX. I guess even that case could be commented - or maybe replaced by a more intuitive `#ifdef __x86_64__` (or the 32 bit equivalent, if anyone cares) to explain what we intend to identify here.

tjhei replied:
My question is why the SSE2 check if necessary. It should also warn if the user compiles deal.II with AVX and the own program without any parallelization (so SSE2 is not defined). I think this should still produce an error, so I would remove ``&& defined(SSE2)`` here.

kronbichler replied:
At least on x86-64, `__SSE2__` should always be defined (well, I am not completely sure about MSVC) even if we have explicitly disabled vectorization by e.g. setting `DEAL_II_HAVE_SSE2=OFF` because it is part of x86-64. But I do not insist on this topic, so let me remove it and remember that there is something here - well, the first one to try a non x86 architecture with 256+ bit vectors will get this error :wink:

tjhei commented on include/deal.II/base/vectorization.h:
```diff
@@ -26,29 +26,29 @@
       "Mismatch in vectorization capabilities: AVX was detected during configuration of deal.II and switched on, but it is apparently not available for the file you are trying to compile at the moment. Check compilation flags controlling the instruction set, such as -march=native."
 #  endif
-#  if DEAL_II_COMPILER_VECTORIZATION_LEVEL >= 3 && defined(__SSE2__) && \
+#  if DEAL_II_VECTORIZATION_WIDTH_IN_BITS >= 512 && defined(__SSE2__) && \
```
same here

tjhei commented on include/deal.II/base/vectorization.h:
```diff
@@ -925,7 +925,7 @@ vectorized_transpose_and_store(const bool                            add_into,
 // set some conflicting compile flags which prevent compilation
 
-#  if DEAL_II_COMPILER_VECTORIZATION_LEVEL >= 3 && defined(__AVX512F__)
+#  if DEAL_II_VECTORIZATION_WIDTH_IN_BITS >= 512 && defined(__AVX512F__)
```
should we remove the ``define`` here and fail instead?

kronbichler replied:
Again, it is to hedge for the non-x86-64 case.

tjhei commented on include/deal.II/base/vectorization.h:
```diff
@@ -2125,7 +2125,7 @@ vectorized_transpose_and_store(const bool                        add_into,
 #  endif
 
-#  if DEAL_II_COMPILER_VECTORIZATION_LEVEL >= 2 && defined(__AVX__)
+#  if DEAL_II_VECTORIZATION_WIDTH_IN_BITS >= 256 && defined(__AVX__)
```
same here and below

tjhei commented on source/base/vectorization.cc:
```diff
@@ -30,31 +30,31 @@ static_assert(std::is_standard_layout<VectorizedArray<float>>::value &&
 
-#if DEAL_II_COMPILER_VECTORIZATION_LEVEL >= 1 && !defined(DEAL_II_MSVC)
-#  if DEAL_II_COMPILER_VECTORIZATION_LEVEL >= 3 && defined(__AVX512F__)
+#if DEAL_II_VECTORIZATION_WIDTH_IN_BITS >= 128 && !defined(DEAL_II_MSVC)
```
Do you know why we exclude MSVC? This should have a comment here.

kronbichler replied:
Some systems require the explicit instantiation of the `static constexpr` variant, others like MSVC do explicitly not want it. Let me find the respective change in the history.

tjhei commented on source/base/vectorization.cc:
```diff
@@ -30,31 +30,31 @@ static_assert(std::is_standard_layout<VectorizedArray<float>>::value &&
-#if DEAL_II_COMPILER_VECTORIZATION_LEVEL >= 1 && !defined(DEAL_II_MSVC)
-#  if DEAL_II_COMPILER_VECTORIZATION_LEVEL >= 3 && defined(__AVX512F__)
+#if DEAL_II_VECTORIZATION_WIDTH_IN_BITS >= 128 && !defined(DEAL_II_MSVC)
+#  if DEAL_II_VECTORIZATION_WIDTH_IN_BITS >= 512 && defined(__AVX512F__)
```
I think non of the ``defined(avx|avx512|...)`` make sense here.

kronbichler replied:
I agree, this should be purely by the available sets in the header file (where the distinction makes sense).

tjhei commented on tests/base/utilities_02b.cc:
```diff
@@ -47,17 +47,17 @@ main()
   initlog();
 
-#if DEAL_II_COMPILER_VECTORIZATION_LEVEL >= 3 && defined(__AVX512F__)
+#if DEAL_II_VECTORIZATION_WIDTH_IN_BITS >= 512 && defined(__AVX512F__)
```
same here and in all other tests: remove the ``defined``

kronbichler commented:
@tjhei I addressed your comments except for the two places mentioned above where I instead added more comments to the code why it looks like that. I also adapted the step-37 and step-48 output and wrote a changelog entry. Should be complete now.

tjhei commented:
Code looks good now. I would propose again to keep the old #define around without deprecation.

kronbichler commented:
I added the two settings as separate commits to be able to identify the respective places slightly more easily.

kronbichler commented:
Any additional comments?

tjhei commented on doc/news/changes/incompatibilities/20200324MartinKronbichler:
```diff
@@ -0,0 +1,5 @@
+Deprecated: The macro DEAL_II_COMPILER_VECTORIZATION_LEVEL (using values 1, 2,
```
can you update this (we don't deprecate after all).

kronbichler replied:
Right, let us move this to minor instead.
